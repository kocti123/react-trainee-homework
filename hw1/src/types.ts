import { AnyAction } from "redux";
import { ThunkAction } from "redux-thunk";

export interface ITask {
  readonly id: number;
  readonly body: string;
  readonly completed: boolean;
  readonly favourite: boolean;
  readonly isEdit: boolean;
  readonly newBody: string;
  readonly creationDate: Date;
}

export type filters = "all" | "completed" | "favourite" | "notCompleted";

export type TasksState = {
  readonly isLoaded: boolean;
  readonly lastId: number;
  readonly tasks: ITask[];
  readonly currentFilter: filters;
  readonly filtredTasks: ITask[];
};

export type State = {
  readonly tasks: TasksState;
};

export type AppAction = ThunkAction<void, State, unknown, AnyAction>;
