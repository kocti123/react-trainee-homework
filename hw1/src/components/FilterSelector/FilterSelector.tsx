import { useAppDispatch, useAppSelector } from "../../redux";
import { changeFilter } from "../../redux/actions/tasksActions";
import { filters } from "../../types";

import styles from "./FilterSelector.module.css";

function FilterSelector() {
  const dispatch = useAppDispatch();
  const { currentFilter } = useAppSelector((store) => store.tasks);

  const filterChangeHandler = (filter: filters) => {
    if (filter === currentFilter) {
      dispatch(changeFilter("all"));
      return;
    }
    dispatch(changeFilter(filter));
  };

  return (
    <div className={styles.filterSelector}>
      <button
        className={currentFilter === "completed" ? styles.selected : undefined}
        onClick={() => filterChangeHandler("completed")}
      >
        Выполненные задачи
      </button>
      <button
        className={
          currentFilter === "notCompleted" ? styles.selected : undefined
        }
        onClick={() => filterChangeHandler("notCompleted")}
      >
        Задачи в работе
      </button>
      <button
        className={currentFilter === "favourite" ? styles.selected : undefined}
        onClick={() => filterChangeHandler("favourite")}
      >
        Избранные задачи
      </button>
    </div>
  );
}

export default FilterSelector;
