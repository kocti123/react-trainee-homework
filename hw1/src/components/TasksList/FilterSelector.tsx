import { useAppDispatch, useAppSelector } from "../../redux";
import { changeFilter } from "../../redux/actions/tasksActions";

import styles from "./FilterSelector.module.css";

function FilterSelector() {
  const dispatch = useAppDispatch();
  const { currentFilter } = useAppSelector((store) => store.tasks);

  const filterFavourite = () => {
    if ("favourite" === currentFilter) {
      dispatch(changeFilter("all"));
      return;
    }
    dispatch(changeFilter("favourite"));
  };
  const filterCompleted = () => {
    if ("completed" === currentFilter) {
      dispatch(changeFilter("all"));
      return;
    }
    dispatch(changeFilter("completed"));
  };
  const filterNotCompleted = () => {
    if ("notCompleted" === currentFilter) {
      dispatch(changeFilter("all"));
      return;
    }
    dispatch(changeFilter("notCompleted"));
  };

  return (
    <div className={styles.filterSelector}>
      <button
        className={currentFilter === "completed" ? styles.selected : undefined}
        onClick={filterCompleted}
      >
        Выполненные задачи
      </button>
      <button
        className={
          currentFilter === "notCompleted" ? styles.selected : undefined
        }
        onClick={filterNotCompleted}
      >
        Задачи в работе
      </button>
      <button
        className={currentFilter === "favourite" ? styles.selected : undefined}
        onClick={filterFavourite}
      >
        Избранные задачи
      </button>
    </div>
  );
}

export default FilterSelector;
