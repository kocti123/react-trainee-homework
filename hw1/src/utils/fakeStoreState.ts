import { State } from "../types";

export function getFakeStoreState(): State {
  return {
    tasks: {
      isLoaded: true,
      lastId: 1,
      currentFilter: "all",
      tasks: [
        {
          id: 1,
          body: "test",
          completed: false,
          favourite: false,
          isEdit: false,
          newBody: "",
          creationDate: new Date(1635696113017),
        },
      ],
      filtredTasks: [
        {
          id: 1,
          body: "test",
          completed: false,
          favourite: false,
          isEdit: false,
          newBody: "",
          creationDate: new Date(1635696113017),
        },
      ],
    },
  };
}
