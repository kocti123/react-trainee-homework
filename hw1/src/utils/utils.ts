import { AnyAction } from "redux";
import { tasksActionTypes } from "../constants/actionTypes";
import { filters, ITask } from "../types";

export const filterTasks = (tasks: ITask[], filter: filters): ITask[] => {
  switch (filter) {
    case "all":
      return tasks;
    case "favourite":
      return filterFavouriteAndNotCompleted(tasks);
    case "completed":
      return filterCompleted(tasks);
    case "notCompleted":
      return filterNotCompleted(tasks);
    default:
      throw new Error("wrong filter type");
  }
};

export const filterFavouriteAndNotCompleted = (tasks: ITask[]): ITask[] => {
  return tasks.filter((task) => task.favourite && !task.completed);
};

export const filterCompleted = (tasks: ITask[]): ITask[] => {
  return tasks.filter((task) => task.completed);
};

export const filterNotCompleted = (tasks: ITask[]): ITask[] => {
  return tasks.filter((task) => !task.completed);
};

export const filterOutById = (tasks: ITask[], id: number): ITask[] => {
  return tasks.filter((task) => task.id !== id);
};

export const findById = (tasks: ITask[], id: number): ITask => {
  return tasks.filter((task) => task.id === id)[0];
};

export const lastIdOfTasks = (tasks: ITask[]): number => {
  return Math.max(...tasks.map((task) => task.id));
};

export const filterAndSortTasks = (
  tasks: ITask[],
  filter: filters
): ITask[] => {
  const filteredTasks = filterTasks(tasks, filter);
  return filteredTasks.sort((a, b) => b.id - a.id);
};

export const getUpdateAction = (id: number, task: ITask): AnyAction => {
  return {
    type: tasksActionTypes.UPDATE_TASK,
    payload: {
      id,
      task: {
        ...task,
      },
    },
  };
};
