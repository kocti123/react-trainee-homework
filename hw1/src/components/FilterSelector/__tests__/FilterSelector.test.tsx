import exp from "constants";
import { mount, ReactWrapper, shallow, ShallowWrapper } from "enzyme";
import { Provider } from "react-redux";
import createMockStore, { MockStoreEnhanced } from "redux-mock-store";
import thunk from "redux-thunk";
import { AppAction, filters, State } from "../../../types";
import { filterTasks } from "../../../utils/utils";
import FilterSelector from "../FilterSelector";
import styles from "../FilterSelector.module.css";

const mockActionType = {
  changeFilter: "changeFilter",
};

jest.mock("../../../redux/actions/tasksActions", () => ({
  changeFilter:
    (filter: filters): AppAction =>
      (dispatch) => {
        dispatch({
          type: mockActionType.changeFilter,
          payload: filter,
        });
      },
}));

const createStore = createMockStore([thunk]);

describe("FilterSelector", () => {
  let wrapper: ReactWrapper;
  let store: MockStoreEnhanced<unknown, {}>;

  beforeEach(() => {
    store = createStore({ tasks: { currentFilter: "all" } });
    jest.spyOn(store, "dispatch");
    wrapper = mount(<FilterSelector />, {
      wrappingComponent: Provider,
      wrappingComponentProps: { store },
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should change filter when click on button", () => {
    wrapper.find("button").forEach((node) => node.simulate("click"));

    expect(store.getActions()).toEqual([
      {
        type: mockActionType.changeFilter,
        payload: "completed",
      },
      {
        type: mockActionType.changeFilter,
        payload: "notCompleted",
      },
      {
        type: mockActionType.changeFilter,
        payload: "favourite",
      },
    ]);
  });

  it("should set class on active filter", () => {
    store = createStore({ tasks: { currentFilter: "completed" } });
    wrapper = mount(<FilterSelector />, {
      wrappingComponent: Provider,
      wrappingComponentProps: { store },
    });
    expect(wrapper.find(`.${styles.selected}`)).toHaveLength(1);

    store = createStore({ tasks: { currentFilter: "notCompleted" } });
    wrapper = mount(<FilterSelector />, {
      wrappingComponent: Provider,
      wrappingComponentProps: { store },
    });
    expect(wrapper.find(`.${styles.selected}`)).toHaveLength(1);

    store = createStore({ tasks: { currentFilter: "favourite" } });
    wrapper = mount(<FilterSelector />, {
      wrappingComponent: Provider,
      wrappingComponentProps: { store },
    });
    expect(wrapper.find(`.${styles.selected}`)).toHaveLength(1);
  });

  it("should disable filter when click on acitve filter", () => {
    store = createStore({ tasks: { currentFilter: "completed" } });
    wrapper = mount(<FilterSelector />, {
      wrappingComponent: Provider,
      wrappingComponentProps: { store },
    });

    wrapper.find(`.${styles.selected}`).simulate("click");

    expect(store.getActions()).toEqual([
      {
        type: mockActionType.changeFilter,
        payload: "all",
      },
    ]);
  });
});
