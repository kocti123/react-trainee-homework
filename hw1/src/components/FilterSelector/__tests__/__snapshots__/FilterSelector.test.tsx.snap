// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSelector should match snapshot 1`] = `ReactWrapper {}`;
